service/src/main/java/org/cbioportal/service/impl/MutationalSignatureServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+package org.cbioportal.service.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.cbioportal.model.GeneticProfile;
+import org.cbioportal.model.MutationalSignature;
+import org.cbioportal.model.SNPCount;
+import org.cbioportal.model.Sample;
+import org.cbioportal.persistence.GeneticProfileRepository;
+import org.cbioportal.persistence.MutationalSignatureRepository;
+import org.cbioportal.persistence.SampleRepository;
+import org.cbioportal.service.MutationalSignatureService;
+
+@Service
+public class MutationalSignatureServiceImpl implements MutationalSignatureService {
+
+	public static class MutationalSignatureFactory {
+		private static final Map<String, String> COMPLEMENTARY_NUCLEOTIDE = new HashMap<>();
+		private static final String[] CANONICAL_SNP_TYPES = new String[]{"C>A", "C>G", "C>T", "T>A", "T>C", "T>G"};
+		private static final Map<String, Integer> CANONICAL_SNP_TYPES_INDEX;
+		
+		static {
+			COMPLEMENTARY_NUCLEOTIDE.put("A","T");
+			COMPLEMENTARY_NUCLEOTIDE.put("T","A");
+			COMPLEMENTARY_NUCLEOTIDE.put("C","G");
+			COMPLEMENTARY_NUCLEOTIDE.put("G","C");
+			
+			CANONICAL_SNP_TYPES_INDEX = new HashMap<>();
+			for (int j = 0; j < CANONICAL_SNP_TYPES.length; j++) {
+				CANONICAL_SNP_TYPES_INDEX.put(CANONICAL_SNP_TYPES[j], j);
+			}
+		}
+		
+		private static Integer getCanonicalSNPIndex(String referenceAllele, String tumorAllele) {
+			if (!referenceAllele.equals("C") && !referenceAllele.equals("T")) {
+				referenceAllele = COMPLEMENTARY_NUCLEOTIDE.get(referenceAllele);
+				tumorAllele = COMPLEMENTARY_NUCLEOTIDE.get(tumorAllele);
+				if (referenceAllele == null || tumorAllele == null) {
+					// Something's wrong
+					return -1;
+				}
+			}
+			Integer index = CANONICAL_SNP_TYPES_INDEX.get(referenceAllele+">"+tumorAllele);
+			if (index == null) {
+				index = -1;
```

Comment:
I didn't understand the use cases. I think I was imagining that our mutation event profiles might involve calls such as A -> Y (where Y is an ambiguous code for T or C). dbSNP can also contain ambiguous nucleotide codes (according to ucsc browser documentation). Maybe when we construct mutation profiles we don't import cases like this ... or don't get them from the platforms which do the calling.